src/charm.py

```diff
@@ -183,14 +182,57 @@ def _check_required_config(self, config_list):
             if self.config[param].strip() == "":
                 raise ValueError(f"Invalid config: {param} value missing")
 
+    def _validate(self, event):  # noqa: C901
+        """Validate that configuration and relations are valid and ready.
+
+        Args:
+            event: The event triggered when the relation changed.
+
+        Raises:
+            ValueError: in case of invalid configuration.
+        """
+        log_level = self.model.config["log-level"].lower()
+        if log_level not in VALID_LOG_LEVELS:
+            raise ValueError(f"config: invalid log level {log_level!r}")
+
+        if not self._state.is_ready():
+            raise ValueError("peer relation not ready")
+
+        self._process_wheel_file(event)
+        self._process_env_file(event)
+
+        self._check_required_config(REQUIRED_CHARM_CONFIG)
+
+        if self._state.supported_workflows is None or len(self._state.supported_workflows) == 0:
+            raise ValueError("Invalid state: must have at least one supported workflow")
+
+        if self._state.supported_activities is None or len(self._state.supported_activities) == 0:
+            raise ValueError("Invalid state: must have at least one supported activity")
+
+        if self._state.module_name is None:
+            raise ValueError("Invalid state: error extracting folder name from wheel file")
+
+        if self.config["auth-enabled"]:
+            if not self.config["auth-provider"]:
+                raise ValueError("Invalid config: auth-provider value missing")
+
+            if not self.config["auth-provider"] in SUPPORTED_AUTH_PROVIDERS:
+                raise ValueError("Invalid config: auth-provider not supported")
+
+            if self.config["auth-provider"] == "candid":
+                self._check_required_config(REQUIRED_CANDID_CONFIG)
+
+            if self.config["auth-provider"] == "google":
```

Comment:
Could be an if else here? Or a switch-case, I know that a thing from a fairly recent version of Python.